server/src/main/java/org/elasticsearch/index/mapper/TextFieldMapper.java

```diff
@@ -161,18 +185,117 @@ public TextFieldMapper build(BuilderContext context) {
+        private final Analyzer delegate;
+
+        PrefixWrappedAnalyzer(Analyzer delegate, int minChars, int maxChars) {
+            super(delegate.getReuseStrategy());
+            this.delegate = delegate;
+            this.minChars = minChars;
+            this.maxChars = maxChars;
+        }
+
+        @Override
+        protected Analyzer getWrappedAnalyzer(String fieldName) {
+            return delegate;
+        }
+
+        @Override
+        protected TokenStreamComponents wrapComponents(String fieldName, TokenStreamComponents components) {
+            TokenFilter filter = new EdgeNGramTokenFilter(components.getTokenStream(), minChars, maxChars);
+            return new TokenStreamComponents(components.getTokenizer(), filter);
+        }
+    }
+
+    private static final class PrefixFieldType extends StringFieldType {
+
+        final int minChars;
+        final int maxChars;
+
+        PrefixFieldType(String name, int minChars, int maxChars) {
+            setTokenized(true);
+            setName(name);
+            this.minChars = minChars;
+            this.maxChars = maxChars;
+        }
+
+        PrefixFieldType setAnalyzer(NamedAnalyzer delegate) {
+            setIndexAnalyzer(new NamedAnalyzer(delegate.name(), AnalyzerScope.INDEX,
+                new PrefixWrappedAnalyzer(delegate.analyzer(), minChars, maxChars)));
+            return this;
+        }
+
+        boolean accept(int length) {
+            return length >= minChars && length <= maxChars;
+        }
+
+        void doXContent(XContentBuilder builder) throws IOException {
+            builder.startObject("index_prefix");
+            builder.field("min_chars", minChars);
+            builder.field("max_chars", maxChars);
+            builder.endObject();
+        }
+
+        @Override
+        public MappedFieldType clone() {
+            return new PrefixFieldType(name(), minChars, maxChars);
+        }
+
+        @Override
+        public String typeName() {
+            return "prefix";
+        }
+
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    private static final class PrefixFieldMapper extends FieldMapper {
+
+        protected PrefixFieldMapper(PrefixFieldType fieldType, Settings indexSettings) {
+            super(fieldType.name(), fieldType, fieldType, indexSettings, MultiFields.empty(), CopyTo.empty());
+        }
+
+        void addField(String value, List<IndexableField> fields) {
+            fields.add(new Field(fieldType().name(), value, fieldType()));
+        }
+
+        @Override
+        protected void parseCreateField(ParseContext context, List<IndexableField> fields) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        protected String contentType() {
+            return "prefix";
+        }
+    }
+
     public static final class TextFieldType extends StringFieldType {
```

Comment:
you will need to implement checkfieldtype so that users get an error if they try to update the `index_prefix` settings

Reply:
And add the modifier in TextFieldTypeTests.